pkg/metricsprovider/prometheus.go

```diff
@@ -0,0 +1,183 @@
+	all_hosts = "all"
+)
+
+func init() {
+	var promHostPresent bool
+
+	promHost, promHostPresent = os.LookupEnv(promHostKey)
+	promToken, promTokenPresent = os.LookupEnv(promTokenKey)
+	if !promHostPresent {
+		promHost = "http://prometheus-k8s:9090"
+	}
+}
+
+type promClient struct {
+	client api.Client
+}
+
+func NewPromClient() (watcher.FetcherClient, error) {
+	var client api.Client
+	var err error
+
+	if !promTokenPresent {
+		client, err = api.NewClient(api.Config{
+			Address: promHost,
+		})
+	} else {
+		client, err = api.NewClient(api.Config{
+			Address: promHost,
+			RoundTripper: config.NewBearerAuthRoundTripper(config.Secret(promToken), api.DefaultRoundTripper),
+		})
+	}
+
+	if err != nil {
+		fmt.Printf("Error creating prometheus client: %v\n", err)
+	}
+
+	return promClient{client}, err
+}
+
+func (s promClient) Name() string {
+	return promClientName
+}
+
+func (s promClient) FetchHostMetrics(host string, window *watcher.Window) ([]watcher.Metric, error) {
+	var metricList []watcher.Metric
+
+	for _, method := range []string{prom_avg_method, prom_std_method} {
+		for _, metric := range []string{prom_cpu_metric, prom_mem_metric} {
+			promQuery := s.buildPromQuery(host, metric, method, window.Duration)
+			promResults := s.getPromRsts(promQuery)
+			curMetricMap := s.promRsts2MetricMap(promResults, metric, method, window.Duration)
+			metricList = append(metricList, curMetricMap[host]...)
+		}
+	}
+
+	return metricList, nil
+}
+
+// Fetch all host metrics for all methods and resource types
+func (s promClient) FetchAllHostsMetrics(window *watcher.Window) (map[string][]watcher.Metric, error) {
+	hostMetrics := make(map[string][]watcher.Metric)
+
+	for _, method := range []string{prom_avg_method, prom_std_method} {
+		for _, metric := range []string{prom_cpu_metric, prom_mem_metric} {
+			promQuery := s.buildPromQuery(all_hosts, metric, method, window.Duration)
+			promResults := s.getPromRsts(promQuery)
+			curMetricMap := s.promRsts2MetricMap(promResults, metric, method, window.Duration)
+
+			for k, v := range curMetricMap {
+				hostMetrics[k] = append(hostMetrics[k], v...)
+			}
+		}
+	}
+
+	return hostMetrics, nil
+}
+
+func (s promClient) buildPromQuery(host string, metric string, method string, rollup string) string {
+	var promQuery string
+	if host == all_hosts {
+		promQuery = fmt.Sprintf("%s(%s[%s])", method, metric, rollup)
+	} else {
+		promQuery = fmt.Sprintf("%s(%s{instance=\"%s\"}[%s])", method, metric, host, rollup)
+	}
+
+	return promQuery
+}
+
+func (s promClient) getPromRsts(promQuery string) model.Value {
+	v1api := v1.NewAPI(s.client)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	results, warnings, err := v1api.Query(ctx, promQuery, time.Now())
+	if err != nil {
+		fmt.Printf("Error querying Prometheus: %v\n", err)
+		os.Exit(1)
+	}
+	if len(warnings) > 0 {
+		fmt.Printf("Warnings: %v\n", warnings)
+	}
+	fmt.Printf("Result:\n%v\n", results.Type())
```

Comment:
Let's move this to debug level when you log to the logger, else the logs will be flooded with query results.